### PR TITLE
Add addrecord command

### PR DESCRIPTION
**Pull request by hjoneweek**

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddRecordCommand.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        List<Person> lastShownList = model.getFilteredPersonList();
+
+        if (index.getZeroBased() >= lastShownList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
+        }
+
+        Person personToAddRecord = lastShownList.get(index.getZeroBased());
+        List<Record> newRecords = new ArrayList<>(personToAddRecord.getRecords());
+        newRecords.add(record);
+        Person personWithAddedRecord = new Person(personToAddRecord.getName(), personToAddRecord.getEmail(),
```
Comment (Darren159) [inline src/main/java/seedu/address/logic/commands/AddRecordCommand.java:59]: Should add a check to see if the existing Record already exists to prevent duplicate Records

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -166,4 +168,43 @@ public static Email parseEmail(String email) throws ParseException {
...
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code condition} is invalid.
+     */
+    public static Condition parseCondition(String condition) {
+        requireNonNull(condition);
+        String trimmedCondition = condition.trim();
+
+        return new Condition(trimmedCondition);
+    }
```
Comment (Darren159) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:194]: Should add checks on whether they are valid DateTime and Condition, can take reference from rest of code.

File: src/main/java/seedu/address/storage/JsonAdaptedCondition.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    public String getCondition() {
+        return condition;
+    }
+    /**
+     * Converts this Jackson-friendly adapted tag object into the model's {@code Condition} object.
+     *
+     * @throws IllegalValueException if there were any data constraints violated in the adapted condition.
+     */
+    public Condition toModelType() {
+        return new Condition(condition);
```
Comment (Darren159) [inline src/main/java/seedu/address/storage/JsonAdaptedCondition.java:41]: Should check if isValidCondition, similar to allergy

File: src/main/java/seedu/address/storage/JsonAdaptedRecord.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import seedu.address.model.record.DateTime;
+import seedu.address.model.record.Record;
+
+/**
+ * Jackson-friendly version of {@link Record}.
+ */
+public class JsonAdaptedRecord {
+
+    public static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("ddMMyyyy HHmm");
+    public static final String MISSING_FIELD_MESSAGE_FORMAT = "Person's %s field is missing!";
```
Comment (Darren159) [inline src/main/java/seedu/address/storage/JsonAdaptedRecord.java:23]: Should be Record not Person

File: src/main/java/seedu/address/storage/JsonAdaptedRecord.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import seedu.address.model.record.Condition;
+import seedu.address.model.record.DateTime;
+import seedu.address.model.record.Record;
+
+/**
+ * Jackson-friendly version of {@link Record}.
+ */
+public class JsonAdaptedRecord {
+
+    public static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("ddMMyyyy HHmm");
```
Comment (Darren159) [inline src/main/java/seedu/address/storage/JsonAdaptedRecord.java:22]: The FORMATTER is not used

File: src/main/java/seedu/address/storage/JsonAdaptedRecord.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+
+    /**
+     * Constructs a {@code JsonAdoptedRecord} with the given record details.
+     */
+    @JsonCreator
+    public JsonAdaptedRecord(@JsonProperty("dateTime") LocalDateTime dateTime,
+                             @JsonProperty("condition") List<JsonAdaptedCondition> conditions) {
+        this.dateTime = dateTime;
+        if (conditions != null) {
```
Comment (Darren159) [inline src/main/java/seedu/address/storage/JsonAdaptedRecord.java:35]: Instantiating a new ArrayList and checking if it is not null is not necessary i think. This means that if the Record condition is null, A JSON with a empty condition is created which shouldnt be the case, it should be an empty Record.

File: src/main/java/seedu/address/storage/JsonAdaptedRecord.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    /**
+     * Converts this Jackson-friendly adapted person object into the model's {@code Person} object.
+     *
+     * @throws IllegalValueException if there were any data constraints violated in the adapted person.
+     */
+    public Record toModelType() throws IllegalValueException {
+        final List<Condition> conditionsList = new ArrayList<>();
+        for (JsonAdaptedCondition condition : conditions) {
+            conditionsList.add(condition.toModelType());
+        }
```
Comment (Darren159) [inline src/main/java/seedu/address/storage/JsonAdaptedRecord.java:59]: This is where the condition is not null check should be done, in the case where the JSON Record is messsed up and there is an empty condition for a Record, then should throw Exception

**Review by Darren159 — COMMENTED** [review]
Overall I think will work but maybe consider using a UniqueRecordList, similar to UniquePersonList so that you can prevent duplicate Records. Need to add more tests also, use CodeCov to find out what tests are missing cause currently its failing the codecov CI test.

**Review by adammangzijun — COMMENTED** [review]
Should be fine, things to take note:
1. More tests needed to cover edge cases.
2. Consider changing list of conditions to a string of description of conditions.
3. Check for valid input of date and time.

File: src/main/java/seedu/address/logic/commands/AddRecordCommand.java (hunk 1 above)
Comment (hjoneweek) [reply src/main/java/seedu/address/logic/commands/AddRecordCommand.java]: Created UniqueRecordList to prevent duplicate records.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 2 above)
Comment (hjoneweek) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java]: Fixed

File: src/main/java/seedu/address/storage/JsonAdaptedRecord.java (hunk 4 above)
Comment (hjoneweek) [reply src/main/java/seedu/address/storage/JsonAdaptedRecord.java]: Fixed

File: src/main/java/seedu/address/storage/JsonAdaptedRecord.java (hunk 5 above)
Comment (hjoneweek) [reply src/main/java/seedu/address/storage/JsonAdaptedRecord.java]: Deleted.

File: src/main/java/seedu/address/storage/JsonAdaptedRecord.java (hunk 6 above)
Comment (hjoneweek) [reply src/main/java/seedu/address/storage/JsonAdaptedRecord.java:33]: Null check is for the the parameter, not the attribute of the JsonAdaptedRecord. Therefore, I think I need the check to pass the test cases.

File: src/main/java/seedu/address/storage/JsonAdaptedRecord.java (hunk 7 above)
Comment (hjoneweek) [reply src/main/java/seedu/address/storage/JsonAdaptedRecord.java]: Fixed.

File: src/main/java/seedu/address/logic/commands/AddRecordCommand.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                && index.equals(otherAddRecordCommand.index);
+    }
+
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this)
+                .add("index", index)
+                .add("record", record)
+                .toString();
+    }
```
[the remark this answers is not in the document]
Comment (hjoneweek) [reply src/main/java/seedu/address/logic/commands/AddRecordCommand.java:92]: Fixed

File: src/main/java/seedu/address/logic/parser/AddRecordCommandParser.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddRecordCommand.MESSAGE_USAGE));
+        }
+
+        Index index = ParserUtil.parseIndex(argMultimap.getPreamble());
+
+        argMultimap.verifyNoDuplicatePrefixesFor(PREFIX_DATE, PREFIX_CONDITION);
+
+        DateTime dateTime = ParserUtil.parseDateTime(argMultimap.getValue(PREFIX_DATE).get());
+        List<Condition> conditions = ParserUtil.parseConditions(argMultimap.getAllValues(PREFIX_CONDITION));
+
```
[the remark this answers is not in the document]
Comment (hjoneweek) [reply src/main/java/seedu/address/logic/parser/AddRecordCommandParser.java:36]: We will leave it as a list for now.

File: src/main/java/seedu/address/model/record/DateTime.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Represents the date and time in which patient visits the doctor
+ */
+public class DateTime {
+    public static final String MESSAGE_CONSTRAINTS =
+            "Date should be in the form of 'dd-MM-yyyy HHmm";
+
+    public static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("ddMMyyyy HHmm");
```
[the remark this answers is not in the document]
Comment (hjoneweek) [reply src/main/java/seedu/address/model/record/DateTime.java:16]: Fixed.

**Review by adammangzijun — APPROVED** [review]
Should be fine, things to take note:
1. More tests needed to cover edge cases.
2. Consider changing list of conditions to a string of description of conditions.
3. Check for valid input of date and time.
